### DISCUSSION TITLE
Parsing PSS/e to PSY

### DISCUSSION THREAD

**Discussion by jd-lara** [Sienna-Platform/PowerSystems.jl#767] [Ideas]
For September:

- [x] Compare PSID/PSSE for September presentation using the same WECC 240 stuff: REGC_A, REEC_B and REPC_A (no PSCAD). [Explanation of why is better to avoid the RECG_A, REEC_B combo in pscad](https://ietresearch.onlinelibrary.wiley.com/doi/epdf/10.1049/iet-gtd.2019.0346)
- [ ] Future implementation should use REGC_C, REEC_D, REPC_A for final comparison PSID/PSSE/PSCAD.

PSS/e models for renewable energy are based on second-generation models, even WT and PV models are compilations of generic ones. 

In a sense, there is a PSS/e meta-model of renewables interfaced through inverters (don't include Type 3 wind turbines) that needs to be translated to [PSY's metamodel](https://nrel-siip.github.io/PowerSystems.jl/stable/modeler_guide/example_dynamic_data/). PSS/e uses an odd terminology: Renewable Exciter, Renewable Governor, Renewable Plant to refer to the blocks of their model. It is unclear what is compatible with what also. 

<img width="1475" alt="image" src="https://user-images.githubusercontent.com/16385323/124317055-8c1e7f00-db44-11eb-96b5-7f76f20e748e.png">

Nominally we have to parse information from the following block diagrams:

- REGCA1: Generic Converter/Machine Model that interfaces with the grid. The final outputs are the current injections. It integrates the converter delays and limits but critically for the PSY metamodel, it doesn't include the AC facing filter.  

![image](https://user-images.githubusercontent.com/16385323/122112497-c6aeba80-cdd5-11eb-86d8-de683cee7866.png)

[ PowerWorld ref](https://www.powerworld.com/WebHelp/Content/TransientModels_HTML/Machine%20Model%20REGC_A.htm)

[Andes ref](https://docs.andes.app/en/latest/groupdoc/RenGen.html#regca1)

- REECB1: PV power a first layer outer control loop that controls the P and Q setpoints. It contains the control loops that go from P and Q references to Iq, Ip references.  PSS/e names these Electrical Controls

![image](https://user-images.githubusercontent.com/16385323/122112662-ec3bc400-cdd5-11eb-80c7-b628bc673329.png)

- REECCU1: Battery storage first layer outer control loop that controls the P and Q setpoints. It contains the control loops that go from P and Q references to Iq, Ip references. This is very similar to the REECB1 except that it has SOC controls. PSS/e names these Electrical Controls

![image](https://user-images.githubusercontent.com/16385323/122112650-e9d96a00-cdd5-11eb-8422-7bf5887e1e39.png)

[PowerWorld ref](https://www.powerworld.com/WebHelp/Content/TransientModels_HTML/Exciter%20REEC_C.htm)

- REPCA1: Generic Renewable second layer outer control that controls V and Frequency set points to generate P and Q set points. PSS/e names these Plant Control Models. From the block diagram, it seems that this control enables a pseudo grid following control (setting P directly) and a frequency droop control model. 

![image](https://user-images.githubusercontent.com/16385323/122111008-17251880-cdd4-11eb-9770-acbb566dbd9a.png)

[PowerWorld ref](https://www.powerworld.com/WebHelp/Content/TransientModels_HTML/Plant%20Controller%20REPC_A.htm)

[Andes ref](https://docs.andes.app/en/latest/groupdoc/RenPlant.html#repca1)

**Notes**: 

1. See [PowerWorld's discussion for reference](https://www.powerworld.com/files/T12RenewableGeneration_2019.pdf)
2. There is also REECA1 that seems to be used for Wind Power mostly and generalizes REECB1. According to the presentation from PowerWorld REECB1 shouldn't be used anymore. Unclear what to do. This is the [implementation in Andes]([Andes ref](https://docs.andes.app/en/latest/groupdoc/RenExciter.html)
3. There is another layer of control for the "DCSource" (acknowledging that the current naming in PSY isn't the best) named Drive Train Models that include PV panel models, aerodynamics models, and torque controllers which we don't currently support.

**Comment by m-bossart** [comment]
A couple of preliminary thoughts. 

1.  In the WECC 240 case, the modelling of wind and solar seems to be identical (same models, same parameters)
2.  There aren't any mechanical models for wind in the WECC 240 case. 
3.  The various flags in the models will change the structure of the model. Is the goal to parse any generic model or first just the ones found in the WECC 240 case (all of the models have the same set of flags).

**Reply by rodrigomha** [reply]
Can you post what are the flags used for the WECC 240 case?

**Reply by rodrigomha** [reply]
Answering myself: In the other sheets of the meta-model separation:
https://docs.google.com/spreadsheets/d/1WnP484Ojm46X6nSLNuYPxGj8hJsfyIhpmD7hZA06Wgg/edit?usp=sharing

**Reply by jd-lara** [reply]
<img width="1586" alt="image" src="https://user-images.githubusercontent.com/16385323/124316413-91c79500-db43-11eb-9ca8-f82c1fd3730a.png">

https://www.wecc.org/Administrative/Wang%20-%20Benchmarking%20between%20EMT%20and%20Positive%20Sequence%20Simulation.pdf

**Comment by m-bossart** [comment]
Initial thoughts on how to fit the inverter meta-model:  https://docs.google.com/spreadsheets/d/1WnP484Ojm46X6nSLNuYPxGj8hJsfyIhpmD7hZA06Wgg/edit?usp=sharing

**Comment by jd-lara** [comment]
Agreement on the separation of REGCB1 and REECA1

![image](https://user-images.githubusercontent.com/16385323/122481901-69a53700-cf84-11eb-94a3-90ec80702ec9.png)

![image](https://user-images.githubusercontent.com/16385323/122482327-2ac3b100-cf85-11eb-98cc-65c8cc70297f.png)

**Reply by rodrigomha** [reply]
There is a minor change in the separation of REECA1 and REECB1 for inner and outer controllers. State 5 should be in the Inner Control, since is a current filtering. The division of Q_ext / V_tfilt has to be done to provide a I_qref for the inner controller, coming from the outer controller. This implies that also in the Active Power part, the division of state 6 by Vt_filt must done in the outer control to provide a I_pref for the inner controller.

**Reply by jd-lara** [reply]
<img width="905" alt="image" src="https://user-images.githubusercontent.com/16385323/124317583-66de4080-db45-11eb-956b-5103da1cd829.png">

**Comment by rodrigomha** [comment]
Discussion is the following:
- REGCA1: It could be Converter(or Filter), It could be an Inner Control. Our approach is to add this one to a Converter (that is current source), and change `Filter` to `GridInterface` (that should transform into the base power of the system).
- REEC_B: It will be divided by InnerControl (Voltage Stuff and Current Limiter) and OuterControl (Power part)
- REPCA: Outer Control that has to be integrated with REEC_B and REEC_A.

**Reply by jd-lara** [reply]
We probably need to move towards a model that uses REGC_C.

**Comment by jd-lara** [comment]
Work on this will continue on the fork https://github.com/Energy-MAC/PowerSystems.jl to open a clean PR in the future with this work.

**Comment by jd-lara** [comment]
For reference, this modeling guideline came out 20 days ago http://www.caiso.com/Documents/InverterBasedInterconnectionRequestsIBRDynamicModelReviewGuideline.pdf#:~:text=General%20model%20description%3A%20REGC%20is%20used%20to%20represent,and%20reactive%20%28Iq%29%20current%20injection%20to%20the%20grid.

**Comment by rodrigomha** [comment]
@jd-lara @m-bossart This is the final separation of the REPCA and REECB models:
Active Renewable Simple Outer Controller:
![activerenewablesimple](https://user-images.githubusercontent.com/24548207/123495885-5f191c00-d5da-11eb-9838-497d44185d6e.jpg)


Reactive Renewable Simple Outer Controller:
![reactiveouterrenewable](https://user-images.githubusercontent.com/24548207/123495673-84595a80-d5d9-11eb-862c-c84a0f5b097c.jpg)

Inner REECB Controller:
![InnerREEC_B](https://user-images.githubusercontent.com/24548207/123495689-963afd80-d5d9-11eb-954e-64e78b6dac21.jpg)

**Comment by jd-lara** [comment]
In the EPRI test guide, we can obtain some reference parameters https://www.wecc.org/Reliability/Test_protocol_REGC_B_REEC_D_EPRI_final.pdf

**Reply by jd-lara** [reply]
RECG_B in PSSe is PLNTBU1

<img width="683" alt="image" src="https://user-images.githubusercontent.com/16385323/124317910-e704a600-db45-11eb-8d3b-162d00247001.png">

**Comment by jd-lara** [comment]
Given that REEC_B will be faced out, we should consider the implementation of REEC_D

![image](https://user-images.githubusercontent.com/16385323/124314727-d271df00-db40-11eb-8dcb-edab9697c00e.png)

**Comment by jd-lara** [comment]
http://cigre-usnc.org/wp-content/uploads/2015/10/CIGRE2015US_Modeling_and_Simulation_of_BESS_for_Grid_Frequency_Regulation.pdf

<img width="1529" alt="image" src="https://user-images.githubusercontent.com/16385323/124315255-b0c52780-db41-11eb-8f0c-d28ec7a95018.png">

**Reply by rodrigomha** [reply]
Pretty good separation, that clearly showcase why we have this issue with the outer and current controller separation. A P/Q control in an inner loop (REEC) is the main issue.

**Comment by jd-lara** [comment]
In a recent conversation about the RECG_A model, we realized that RECG_A is not a current injection source but rather a model of a current source behind an impedance. This means that the RECG_A is, in reality, a voltage source modeled as a Norton Equivalent

![image](https://user-images.githubusercontent.com/16385323/126019238-61b25046-6a3e-44a9-9723-e52d7a53e29d.png)

PSS/e uses the static data from the Source fields from the power flow data to populate this value. 

![image](https://user-images.githubusercontent.com/16385323/126019268-a7966857-8aea-49e5-9848-c1dd91c1cdf0.png)


As a result, we can integrate Z_Source into the converter model and maintain that the converter outputs remain to be voltages.
